### DISCUSSION TITLE
Flattening Nested Properties with In-Memory ProjectTo causes NRE

### DISCUSSION THREAD

**Discussion by richardlawley** [LuckyPennySoftware/AutoMapper#3691] [NA]
I use `ProjectTo` a lot for my EF6 code, but the same code runs with in-memory Queryables for the purpose of Unit Tests.  Previously with 10.0.0 I had to add explicit conditions when flattening a nullable second-level property into a DTO - since 10.1.1 this has changed again so it is also needed with Non-Nullable properties.  Normal `Map<T>` calls handle this without any configuration.

The main problem with this is that my profiles need contain mapping instructions which are only needed for in-memory unit tests, which is obviously undesirable.

[Gist](https://gist.github.com/richardlawley/a0c3bd607c35e1dbf9c25824ff1b61db)

### Source/destination types

```csharp
public class Parent
{
	public int NonNullable { get; set; }
	public int? Nullable { get; set; }
}

public class Item
{
	public Parent Parent { get; set; }   // Optional Field
}

public class ItemDto
{
	public int? ParentNonNullable { get; set; }
	public int? ParentNullable { get; set; }
}
```

### Mapping configuration

```csharp
var config = new MapperConfiguration(c =>
{
	c.CreateMap<Item, ItemDto>()
		// needed to avoid NRE in 10.0.0
		.ForMember(dest => dest.ParentNullable, opt => opt.MapFrom(src => src.Parent != null ? src.Parent.Nullable : null))

		// needed to avoid NRE in 10.1.1
		//.ForMember(dest => dest.ParentNonNullable, opt => opt.MapFrom(src => src.Parent != null ? (int?)src.Parent.NonNullable : null))
		;
});
```

### Version: 10.1.1

(on net472).  Also tested with latest myget version.
 
### Expected behavior

Properties from `Parent` flatten correctly to nullable destination properties without explicit configuration.  In 10.0.0, this worked for Nullable nested properties.

### Actual behavior

NRE if Parent is null and no explicit configuration

### Steps to reproduce

```csharp
var data = new List<Item>
{
	new Item()		// Note Parent is null
};

var result = data.AsQueryable()
	.ProjectTo<ItemDto>(mapper.ConfigurationProvider)
	.ToList();
```

**Comment by lbargaoanu** [comment]
`ProjectTo` in memory is not relevant. You can use an in memory DB instead.

**Reply by L45eMy** [reply]
It's a bit strange that the behavior got changed in a minor update. Projection might be used on other sources that implement IQueryable, not only database providers that handle the null reference itself.

**Reply by jbogard** [reply]
We don't design ProjectTo for anything else but database providers. Even past that, LINQ support varies TREMENDOUSLY amongst different DB providers, so we still just try to target EF and EF Core. Mongo has one, Cosmos has one, RavenDB has one etc etc.
